File: app/src/main/java/fr/paug/androidmakers/ui/adapter/AgendaPagerAdapter.java
package fr.paug.androidmakers.ui.adapter;

import android.app.Activity;
import android.support.v4.view.PagerAdapter;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.LinearSmoothScroller;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import fr.paug.androidmakers.R;
import fr.paug.androidmakers.ui.activity.SessionDetailActivity;
import fr.paug.androidmakers.ui.util.SessionFilter;
import fr.paug.androidmakers.util.sticky_headers.StickyHeadersLinearLayoutManager;

public class AgendaPagerAdapter extends PagerAdapter {

    private final List<DaySchedule> mAgenda;
    private final SparseArray<View> mAgendaViews = new SparseArray<>();
    private Activity activity;

    private RecyclerView recyclerView;
    private List<SessionFilter> sessionFilterList = new ArrayList<>();

    public AgendaPagerAdapter(List<DaySchedule> mAgenda, Activity activity) {
        this.mAgenda = mAgenda;
        this.activity = activity;
    }

    public void refreshSessionsSelected() {
        for (int i = 0; i < mAgendaViews.size(); i++) {
            View view = mAgendaViews.get(mAgendaViews.keyAt(i));
            RecyclerView recyclerView = view.findViewById(android.R.id.list);
            recyclerView.getAdapter().notifyDataSetChanged();
        }
    }

    @Override
    public Object instantiateItem(ViewGroup collection, int position) {
        View view = LayoutInflater.from(collection.getContext()).inflate(R.layout.schedule_single_day_list, null, false);

        recyclerView = view.findViewById(android.R.id.list);
        recyclerView.setHasFixedSize(true);
        recyclerView.addItemDecoration(new DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL));
        recyclerView.setItemAnimator(new DefaultItemAnimator());

        recyclerView.setLayoutManager(new StickyHeadersLinearLayoutManager<ScheduleDayAdapter>(activity));

        ScheduleDayAdapter adapter = new ScheduleDayAdapter(activity, getItems(position), true, new ScheduleDayAdapter.OnItemClickListener() {
            @Override public void onItemClick(ScheduleSession agendaItem) {
                SessionDetailActivity.startActivity(activity, agendaItem);
            }
        });

        adapter.setSessionFilterList(sessionFilterList);
        recyclerView.setAdapter(adapter);
        moveToCurrentTimeSlot(true, adapter);

        collection.addView(view);
        mAgendaViews.put(position, view);
        return view;
    }

    @Override
    public void destroyItem(ViewGroup collection, int position, Object view) {
        collection.removeView((View) view);
        mAgendaViews.remove(position);
    }

    @Override
    public int getCount() {
        return mAgenda.size();
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return getItems(position).getTitle();
    }

    private DaySchedule getItems(int position) {
        return mAgenda.get(position);
    }

    private void moveToCurrentTimeSlot(boolean animate, ScheduleDayAdapter adapter) {
        final long now = System.currentTimeMillis(); // current time
        final int position = adapter.findTimeHeaderPositionForTime(now);

        RecyclerView.SmoothScroller smoothScroller = new LinearSmoothScroller(activity) {
            @Override protected int getVerticalSnapPreference() {
                return LinearSmoothScroller.SNAP_TO_START;
            }
        };

        LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();

        if (position >= 0) {
            if (animate) {
                smoothScroller.setTargetPosition(position);
                layoutManager.startSmoothScroll(smoothScroller);
            } else {
                layoutManager.scrollToPositionWithOffset(position,
                        this.activity.getResources().getDimensionPixelSize(R.dimen.default_padding));
            }
        }
    }

    public void setSessionFilterList(List<SessionFilter> sessionFilterList) {
        this.sessionFilterList = sessionFilterList;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/fr/paug/androidmakers/ui/adapter/AgendaPagerAdapter.kt
package fr.paug.androidmakers.ui.adapter

import android.app.Activity
import android.content.Context
import android.util.SparseArray
import android.view.View
import android.view.ViewGroup
import androidx.compose.ui.platform.ComposeView
import androidx.core.util.forEach
import androidx.recyclerview.widget.*
import androidx.viewpager.widget.PagerAdapter
import fr.paug.androidmakers.R
import fr.paug.androidmakers.ui.activity.SessionDetailActivity
import fr.paug.androidmakers.ui.components.AgendaColumn
import fr.paug.androidmakers.ui.model.UISession
import fr.paug.androidmakers.ui.theme.AndroidMakersTheme
import fr.paug.androidmakers.ui.util.SessionFilter
import fr.paug.androidmakers.util.SessionSelector
import fr.paug.androidmakers.util.TimeUtils
import java.time.OffsetDateTime
import java.util.*

class AgendaPagerAdapter(private val mAgenda: List<DaySchedule>, private val activity: Activity) : PagerAdapter() {
    private val mAgendaViews = SparseArray<View>()

    private var sessionFilterList: List<SessionFilter> = ArrayList()

    fun refreshSessionsSelected() {
    }

    private fun addSeparators(context: Context, sessions: List<UISession>): Map<String, List<UISession>> {
        return sessions.map {
            TimeUtils.formatShortTime(context, Date(it.startDate.toEpochMilli())) to it
        }
            .groupBy(
                keySelector =  { it.first }
            ) {
                it.second
            }
    }

    fun getRoomTitle(scheduleSession: ScheduleSession, daySchedule: DaySchedule): String {
        var roomTitle = ""
        for (roomSchedule in daySchedule.roomSchedules) {
            if (roomSchedule.roomId == scheduleSession.roomId) {
                roomTitle = roomSchedule.title
            }
        }
        return roomTitle
    }

    override fun instantiateItem(collection: ViewGroup, position: Int): Any {
        val view = ComposeView(collection.context)

        updateSessions(view, position)

        collection.addView(view)
        mAgendaViews.put(position, view)
        return view
    }

    override fun destroyItem(collection: ViewGroup, position: Int, view: Any) {
        collection.removeView(view as View)
        mAgendaViews.remove(position)
    }

    override fun getCount(): Int {
        return mAgenda.size
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return getItems(position).title
    }

    private fun getItems(position: Int): DaySchedule {
        return mAgenda[position]
    }

    private fun List<ScheduleSession>.filterSessions(filterList: List<SessionFilter>): List<ScheduleSession> {
        val filteredSessions = mutableListOf<ScheduleSession>()

        if (sessionFilterList.isEmpty()) {
            filteredSessions.addAll(this)
        } else {
            for (item in this) {
                for (sessionFilter in sessionFilterList) {
                    val matched = when (sessionFilter.type) {
                        SessionFilter.FilterType.BOOKMARK -> {
                            SessionSelector.isSelected(item.sessionId)
                        }
                        SessionFilter.FilterType.LANGUAGE -> {
                            sessionFilter.value == item.language
                        }
                        SessionFilter.FilterType.ROOM -> {
                            sessionFilter.value == item.roomId
                        }
                    }

                    if (matched) {
                        filteredSessions.add(item)
                    }
                }
            }
        }

        return filteredSessions
    }

    private fun updateSessions(composeView: ComposeView, position: Int) {
        val daySchedule = getItems(position)
        val items = daySchedule.roomSchedules.flatMap { it.scheduleSessions }
            .filterSessions(sessionFilterList)
            .sorted()
            .map { item ->
            UISession(
                id = item.sessionId,
                title = item.title,
                language = item.language,
                startDate = OffsetDateTime.parse(item.slot.startDate).toInstant(),
                endDate = OffsetDateTime.parse(item.slot.endDate).toInstant(),
                room = getRoomTitle(item, daySchedule),
                roomId = item.roomId,
                speakers = item.speakers.map {
                    UISession.Speaker(it.name ?: "")
                },
            )
        }

        val grouped = addSeparators(
            composeView.context,
            items,
        )
        composeView.setContent {
            AndroidMakersTheme {
                AgendaColumn(grouped) {
                    SessionDetailActivity.startActivity(activity, it)
                }
            }
        }
    }

    fun setSessionFilterList(sessionFilterList: List<SessionFilter>) {
        this.sessionFilterList = sessionFilterList

        mAgendaViews.forEach { position, view ->
            updateSessions(view as ComposeView, position)
        }
    }
}